### ISSUE TITLE
Default parameter doesn't work nice with parameter destruturing + rest operator.

### ISSUE THREAD

**Issue by vilic** [evanw/esbuild#1826]
Input:

```js
function foo({a, ...x}, b = a) {}
```

Output:

```js
function foo(_a, b = a) {
  var _b = _a, { a } = _b, x = __objRest(_b, ["a"]);
}
```

```
ReferenceError: a is not defined
```

**Comment by evanw** [comment]
FWIW this is also how [the TypeScript compiler (which esbuild is modeled after) compiles this code](https://www.typescriptlang.org/play?ssl=2&ssc=1&pln=3&pc=1#code/GYVwdgxgLglg9mABMOcAUBvAhgGkQOkIA8BfPAI0QF5EsBKRDAKBKaA). Here is their output:

```js
function foo(_a, b = a) {
    var { a } = _a, x = __rest(_a, ["a"]);
}
```

If they have had this problem this whole time and no one has reported it (I couldn't find an issue), then this probably isn't super urgent to fix. How did this come up?

Note to self: a comment from a while ago talks about this in the code. It's saying there's some trade-off here between evaluation order and scope, although I don't have full context on this at the moment:

https://github.com/evanw/esbuild/blob/746faa06112785c7b7839aeda46b329c20cf0472/internal/js_parser/js_parser_lower.go#L257-L286

**Comment by rtsao** [comment]
The original function throws an error when executed so I don't this is an issue with esbuild.

**Comment by vilic** [comment]
@evanw That is really interesting to know because I have been using TypeScript for years and this is the first time I use esbuild (via vite) for a relatively large project.
@rtsao You probably forgot to provide an object as the first parameter?
<img width="250" alt="image" src="https://user-images.githubusercontent.com/970430/145287472-178f38cf-c047-4381-a49a-03e36bbdc35c.png">

**Comment by vilic** [comment]
Oh I see why, because for Node.js we use tsc and targets es2018, and for browser we were using babel, and babel does it right.

```js
function foo(_ref) {
  var a = _ref.a,
      x = _objectWithoutProperties(_ref, _excluded);

  var b = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : a;
}
```

Not that urgent for me, but I can think of the potential impact on introducing an extra build step if we need more compatibility in the future.

**Comment by rtsao** [comment]
Yeah, I would suspect a sizable percentage of TS users don't actually rely on tsc for syntax lowering and are using Babel instead. It's certainly plausible it hasn't come up all this time for that reason...

**Comment by adrienharnay** [comment]
Hi @evanw, TS seems to has fixed the issue on their end. Do you think we could backport the fix to esbuild? I've been bitten by this today 😬
